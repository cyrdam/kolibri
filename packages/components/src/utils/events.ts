// TODO: Should be synchronized with enums/events.ts
export type KoliBriEventType =
	| 'blur'
	| 'change'
	| 'changePage'
	| 'changePageSize'
	| 'click'
	| 'close'
	| 'focus'
	| 'input'
	| 'mousedown'
	| 'reset'
	| 'selection-change' // todo remove
	| 'selectionChange'
	| 'sort'
	| 'submit'
	| 'toggle';

export function stopPropagation(event: Event): void {
	event.stopImmediatePropagation();
	event.stopPropagation();
}

function createKoliBriEvent<T>(type: KoliBriEventType, detail: T): CustomEvent {
	const event = new CustomEvent(`kol-${type}`, {
		bubbles: true,
		cancelable: true,
		composed: true,
		detail: detail,
	});
	return event;
}

function dispatchKoliBriEvent<T>(target: EventTarget, type: KoliBriEventType, detail?: T): boolean {
	const dispatch = target.dispatchEvent(createKoliBriEvent(type, detail));
	return dispatch;
}

/**
 * @deprecated
 */
export function tryToDispatchKoliBriEvent<T>(type: KoliBriEventType, target?: EventTarget, detail?: T): void {
	target && dispatchKoliBriEvent(target, type, detail);
}

export function dispatchDomEvent(target: HTMLElement, type: KoliBriEventType) {
	target.dispatchEvent(new Event(type, { bubbles: true, composed: true }));
}

export function preventDefaultAndStopPropagation(event: Event) {
	event.preventDefault();
	stopPropagation(event);
}
