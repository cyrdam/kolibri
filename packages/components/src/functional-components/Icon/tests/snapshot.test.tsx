import { h } from '@stencil/core';
import { renderFunctionalComponentToSpecPage } from '../../../utils/testing';
import KolIconFc from '../Icon';

describe('KolIconFc', () => {
	it('should render correctly', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolIconFc icons="test" label="test" />);
		expect(page.root).toMatchSnapshot();
	});

	it('should render with the correct class', async () => {
		const className = 'test-class';
		const page = await renderFunctionalComponentToSpecPage(() => <KolIconFc icons="test" label="test" class={className} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.className).toContain(className);
	});

	it('should render with the correct style', async () => {
		const style = { color: 'red' };
		const page = await renderFunctionalComponentToSpecPage(() => <KolIconFc icons="test" label="test" style={style} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.style.color).toBe('red');
	});

	it('should handle onClick event', async () => {
		const onClick = jest.fn();
		const page = await renderFunctionalComponentToSpecPage(() => <KolIconFc icons="test" label="test" onClick={onClick} />);
		page.root?.click();
		expect(onClick).toHaveBeenCalled();
	});

	it('should render with the correct icons and label', async () => {
		const icons = 'test-icons';
		const label = 'test-label';
		const page = await renderFunctionalComponentToSpecPage(() => <KolIconFc icons={icons} label={label} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.getAttribute('_icons')).toBe(icons);
		expect(page.root?.getAttribute('_label')).toBe(label);
	});
});
