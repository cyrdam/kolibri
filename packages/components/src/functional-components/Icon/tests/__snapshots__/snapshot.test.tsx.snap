// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KolIconFc should render correctly 1`] = `
<kol-icon
  _icons="test"
  _label="test"
/>
`;

exports[`KolIconFc should render with the correct class 1`] = `
<kol-icon
  _icons="test"
  _label="test"
  class="test-class"
/>
`;

exports[`KolIconFc should render with the correct icons and label 1`] = `
<kol-icon
  _icons="test-icons"
  _label="test-label"
/>
`;

exports[`KolIconFc should render with the correct style 1`] = `
<kol-icon
  _icons="test"
  _label="test"
/>
`;
